### ISSUE TITLE
Iteration

### ISSUE THREAD

**Issue by espenfl** [easybuilders/easybuild-framework#2865]
Would it make sense to introduce something like a `easywrapper` that opens for rather easy configuration and iteration of a build matrix, for instance specified using a `yaml`/`json` file? Right now I can not find direct support for iteration, which would be very useful. In our case we would typically generate between 20-50 different versions of the base software (different combination of patches, libraries, -D options etc.) given a fixed toolchain.

**Comment by boegel** [comment]
If I get this right, there would be no need for a wrapper at all, it's just a different (additional) input to the `eb` command.

For example, it could be something like "`eb base.eb --matrix example.yaml`", where `base.eb` is the base easyconfig file and `example.yaml` specifies the build matrix, i.e. the changes to be made on top of the specified easyconfig file.

Underneath the hood `eb` could then fairly easily generate a directory of easyconfig files based on that, and then continue to perform all those installations.

If you would like to give this a shot, the `start` hook mentioned in #2864 could be a good starting point, but some minor changes would need to be made to give you the needed flexibility (you can't tweak the list of easyconfig files that will be processed from the `start` hook right now, but making that possible is fairly easy).

**Comment by espenfl** [comment]
Thanks again for the quick reply. Yes, something like this would be great. Then in the `yaml` we could supply the existing parameters or additional from a custom easyblock (this is probably something I would need, as I then would apply patches with `apply_patch` or modify the `-D` options etc.).

I will have a look at the `start` hook tomorrow. Your idea would then be to use that hook to create new directories etc. and then pass along whatever needs passing to the easyblock?

**Comment by espenfl** [comment]
Okey, so gave this a look now and I think this would be the strategy: since `build_and_install_software` already accepts a list of easyconfigs, the `ecs`, the only thing we really need to do is prepare those files (or objects). This we can do in the `start_hook`. It would be most convenient to generate the easyconfigs as objects. Seems `ecs` is basically a list of dictionaries containing the parameters of the easyconfig (after inspecting `process_easyconfig`). I am sure we could generate this rather easily in the `start_hook`. However, it is not clear to me yet what would be the best strategy to inject this into the downstream processing of EB. Would be nice not having to generate explicit files. Sounds okey? What do you think about the injection?

**Comment by espenfl** [comment]
Also, if avoiding files I think, in `main`, that the assignment of `easyconfigs` from the call to `parse_easyconfigs` need to append to `easyconfigs` if we have entries present from injection in the `start_hook`. It would also be natural to open for such behavior in the hooks in general I think. Not sure about how to handle the flag `generated_ecs` in this context though.

**Comment by espenfl** [comment]
Also, can I do `pip install -e .` after a clone of the `easybuild-framework`? Want to fork and modify the code a bit to try these things.

**Comment by espenfl** [comment]
But is seems `self` is not available in `start_hook`. How then to takes definitions done here downstream in EB?

**Comment by boegel** [comment]
@espenfl That's what I mentioned before, you'll need to modify how the `start_hook` is called currently, so you get a handle to inject easyconfig files before they get processed.

I'm not sure how well it would work out if you avoid generating actual files, since that's not what the logic that is there now expects... But you can definitely give it a shot, any issues that pop up shouldn't be impossible to fix.

**Comment by espenfl** [comment]
@boegel Yes, right. So this in fact needs some modifications to the source code. We put together something that works by simply generating the files and then iterating over the `main` method. We will have a look at how this can be embedded when time permits. Would be nice to just pass along a dictionary, or even a list of dictionaries that you could generate in a json and have it just read that and iterate. Thanks for the help so far.
